app/views/foreman/job_templates/change_content_source.erb

```diff
@@ -0,0 +1,86 @@
+<%#
+kind: job_template
+name: Change content source
+job_category: Katello
+model: JobTemplate
+provider_type: SSH
+description_format: Change host's content source
+feature: katello_change_content_source
+template_inputs:
+- name: api_user
+  required: true
+  input_type: user
+  value_type: plain
+  description: Username of user for API callback
+  advanced: false
+- name: api_user_password
+  required: true
+  input_type: user
+  value_type: plain
+  description: User password for API callback
+  hidden_value: true
+  advanced: false
+- name: smart_proxy
+  required: true
+  input_type: user
+  value_type: resource
+  resource_type: SmartProxy
+  description: New content source (Smart Proxy)
+  advanced: false
+- name: hostgroup
+  required: false
+  input_type: user
+  value_type: resource
+  resource_type: Hostgroup
+  description: Host group for host
+  advanced: false
+%>
+#!/bin/sh
+<%
+    smart_proxy = input_resource('smart_proxy')
+    rhsm_url = smart_proxy.rhsm_url
+    pulp_content_url = smart_proxy.pulp_content_url
+    host_params = {
+        host: {
+            hostgroup_id: (input_resource('hostgroup').id if input_resource('hostgroup')),
+            content_facet_attributes: {
+              content_source_id: smart_proxy.id
+            },
+            content_source_id: smart_proxy.id
+        }
+    }
+-%>
+
+SSL_CA_CERT=$(mktemp)
+cat << EOF > $SSL_CA_CERT
+<%= foreman_server_ca_cert %>
+EOF
+
+
+KATELLO_SERVER_CA_CERT=/etc/rhsm/ca/katello-server-ca.pem
+RHSM_CFG=/etc/rhsm/rhsm.conf
+
+# Prepare SSL certificate
+mkdir -p /etc/rhsm/ca
+cp -f $SSL_CA_CERT $KATELLO_SERVER_CA_CERT
+chmod 644 $KATELLO_SERVER_CA_CERT
+
+# Configure subscription-manager
+test -f $RHSM_CFG.bak || cp $RHSM_CFG $RHSM_CFG.bak
+
+subscription-manager config \
+  --server.hostname="<%= rhsm_url.host %>" \
+  --server.port="<%= rhsm_url.port %>" \
+  --server.prefix="<%= rhsm_url.path %>" \
+  --rhsm.repo_ca_cert="$KATELLO_SERVER_CA_CERT" \
+  --rhsm.baseurl="<%= pulp_content_url %>"
+
+# Update host
```

Comment:
Why is this needed? I would expect the workflow to be:
* User changes content source on the host
* Runs REX job to match the config on the host to what's in Foreman

Essentially: treat Foreman's DB as the source of truth.

Reply:
The primary idea is as you call out to make use of REX to enable changing the proxy a host is pointed at. The secondary idea is to use this same mechanism to allow changing the proxy from the host itself as users do with `bootstrap.py` since we have not achieved REX everywhere across installations. The workflow would be:

 * User on a host hits API to change source
 * Template is rendered and delivered to the host and ran
 
I imagine this being used by an Ansible module or a CLI tool like the newer RHC RHEL is working to provide.

Reply:
> The primary idea is as you call out to make use of REX to enable changing the proxy a host is pointed at.

I'd be against that if I'm honest. I've always seen Foreman as the source of truth. To me it would make more sense to have a "trigger" that if you change the content source on a host that a REX job can automatically be triggered to update the host.

I also see security concerns. Curl runs on the host and there the username and password will be visible in the process tree (`ps aux`). I hadn't mentioned it inline since I first want to have the higher level discussion.



> The workflow would be:
> 
>     * User on a host hits API to change source
> 
>     * Template is rendered and delivered to the host and ran

This makes sense to me and what I was arguing for: treat Foreman as the source of truth and then rectify hosts to match.

Reply:
> * User on a host hits API to change source
> * Template is rendered and delivered to the host and ran

That was original plan, and done in https://github.com/Katello/katello/pull/9713, but then it was decided to go with REX template - it's more customizable and there is no need to maintain another API endpoint

Reply:
In that PR I also see a HTTP call to Foreman to modify data. That is still my problem with this. I think this REX template should be reduced to reading the host property and applying that info to the host in practice. So essentially this REX template without any job parameters and without this curl request.

Reply:
> I also see security concerns.

We could fix it fairly easy. Create new template macro `user_api_auth_token`, that would generate API auth token for the current user. Of course it would allow to set the scope and expiration time, same way as we do in the registration template [0]. 
This way we can call Foreman API without exposing user's credentials. Thoughts?

[0] https://github.com/theforeman/foreman/blob/develop/app/controllers/concerns/foreman/controller/registration_commands.rb#L39

Reply:
It still leaves a window of attack and I don't see the need to. Going Foreman -> Host -> Foreman feels plain wrong to me.